text/0121-data-types.md

```diff
@@ -0,0 +1,450 @@
+
+The following input:
+
+```json
+[
+    [
+        "Bruce Willis",
+        "Vin Diesel"
+    ],
+    "Kung Fu Panda"
+]
+```
+
+Will be processed as if all elements were arranged at the same level:
+
+```json
+"Bruce Willis. Vin Diesel. Kung Fu Panda."
+```
+
+The strings above will be separated by soft and hard spaces exactly as explained in the string example.
+
+> See [3.1.1. Examples section](#311-examples).
+
+### 3.3. Object
+
+#### 3.3.1. Example
+
+The following input:
+
+```json
+{
+    "movie_id": "1564saqw12ss",
+    "title": "Kung fu Panda"
+}
+```
+
+In the example above, `movie_id`, `1564saqw12ss`, `title`, `Kung fu Panda` are all considered sentences. The colon `:` and comma `,` characters are used as separators.
+
+`"movie_id. 1564saqw12ss. title. Kung fu Panda."`
+
+These sentences will be separated by soft and hard spaces precisely as explained in the string example.
+
+### 3.4. Nested Structures
+
+Nested structures (e.g. `Object`, `Array of Objects`, etc) are internally flattened to a document's root level.
+
+It allows expressing a nested field in all Meilisearch parameters that accept document attributes.
+
+Meilisearch accepts the `.` (dot-notation) to express a nested field location in a document structure.
+
+### 3.4.1. Examples
+
+#### 3.4.1.1. Object
+
+The following JSON document:
+
+```json
+    {
+        "a": {
+            "b": "c",
+            "d": "e",
+            "f": "g"
+        }
+    }
+```
+
+Flattens to:
+
+```json
+{
+    "a.b": "c",
+    "a.d": "e",
+    "a.f": "g"
+}
+```
+
+#### 3.4.1.2. Array of objects
+
+The following JSON document:
+
+```json
+{
+    "a": [
+        { "b": "c" },
+        { "b": "d" },
+        { "b": "e" },
+    ]
+}
+```
+
+Flattens to:
+
+```json
+{
+    "a.b": ["c", "d", "e"],
+}
+```
+
+#### 3.4.1.3. Array of objects mixed with scalar value
+
+The following JSON document:
+
+```json
+{
+    "a": [
+        42,
+        { "b": "c" },
+        { "b": "d" },
+        { "b": "e" },
+    ]
+}
+```
+
+Flattens to:
+
+```json
+{
+    "a": 42,
+    "a.b": ["c", "d", "e"],
+}
+```
+
+
+#### 3.4.1.4. Array of objects of array of objects of ...
+
+The following JSON document:
+
+```json
+{
+    "a": [
+        "b",
+        [
+            "c",
+            "d"
+        ],
+        {
+            "e": [
+                "f",
+                "g"
+            ]
+        },
+        [
+            {
+                "h": "i"
+            },
+            {
+                "e": [
+                    "j",
+                    {
+                        "z": "y"
+                    }
+                ]
+            },
+        ],
+        ["l"],
+        "m"
+    ]
+}
+```
+
+Flattens to:
+
+```json
+{
+    "a": ["b", "c", "d", "l", "m"],
+    "a.e": ["f", "g", "j"],
+    "a.h": "i",
+    "a.e.z": "y"
+}
+```
+
+#### 3.2.4.5. Collision between a representation
+
+The following JSON document:
+
+```json
+{
+    "a": {
+        "b": "c",
+    },
+    "a.b": "d"
+}```
```

Comment:
I think there is an error on the formatting. 
```suggestion
}
```
```

Reply:
Thanks @qdequele. 

This part needs to be decided and precise.

e.g. ["person"] instead of ["person.firstname", "person.lastname"].

This notation could easily work on some parameters while not on others.

e.g. attributesToHighlight (works-well) vs searchableAttributes (impossible, it introduces equals field weight for a subset of document fields).

We could first detect this expression on some parameters and block it with an error message, or simply ignore it and specify that each field must be expressly defined with the notation `.` for a first iteration.